### ISSUE TITLE
Sympa CPU activity 

### ISSUE THREAD

**Issue by dpc22** [sympa-community/sympa#1238]
Version
-------

6.2.66

Installation method
-------------------

My own rpm, derived from "official" Redhat 7 RPM

Expected behavior
-----------------

sympa_msg.pl and task_manager.pl should be idle much of the time unless there are new messages or tasks to process.

Actual behavior
---------------

The two daemons seem to spend an alarming amount of time running at 100% user CPU. (sympa_msg.pl also uses an alarming amount of memory at 1.2 Gbytes RSS, but that will be a separate ticket).

Additional information
----------------------

We are migrating 7600 lists from Mailman to Sympa. Currently 4300 down, 3300 to go. I am concerned that sympa_msg.pl and task_manager.pl spend quite a lot of time running at 100% user CPU, given that we still have 3300 lists to go.

A bit of poking around revealed two immediate causes of this load:

1) lib/Sympa/Spool/Task.pm _create_all_tasks() is called every 60 seconds. This runs:
 ```
    foreach my $robot (Sympa::List::get_robots()) {
        my $all_lists = Sympa::List::get_lists($robot);
```
which appears to take 20 to 30 seconds to complete, every 60 seconds.

2) /usr/sbin/sympa_msg.pl processes the digest spool every 25 seconds:

```
    my $digestsleep = 5;
   ...
    # Sleep for a while if spool is empty.
    sleep $Conf::Conf{'sleep'};  # Defaults to 5 seconds
```

That appears to iterate through 1865 lists which _may_ have digest messages, creating a Sympa::List object in each case:
```
$ ls /var/spool/sympa/digest/ | wc -l
1865
```
Only 1642 actual messages queued for daily digests at the time of writing, mid afternoon:
```
$ ls /var/spool/sympa/digest/*/*,* | wc -l
1642
```
which be all be sent out at 04:00 tomorrow morning.

My question: do these two things actually need to run once or twice per minute?

I'm am particularly tempted to work out a way to process the digest spool once a day at 04:00 and just hide the per list time of day 
and digest_max_size configuration from list owners. Only two out of our 4400 Sympa lists have changed the digest setting from the default:

digest 0,1,2,3,4,5,6 4:0

and they provide digests on weekdays only and once a week respectively, both at 04:00:

digest 1,2,3,4 4:0
digest 1 4:0

No list owner has changed the default time of day, and no list owner has changed digest_max_size from the default value of 25. I would be surprised if anyone was actually hitting that limit.

**Comment by racke** [comment]
It would be also good if you could disable the digest entirely.

**Comment by ikedas** [comment]
IMO periodic jobs like tasks or digest sending would be better to be left to external mechanisms such as `crontab`.  However our ancestors decided to implement their own mechanism to control periodic jobs.

**Comment by dpc22** [comment]
I agree that these things should be run from crontab (that's the way that Mailman does it). However would anything bad happen if I increased various existing delays so that tasks and digests only run  once an hour rather than once or twice a minute?

In particular: what happens if a Sympa::Spindle::ProcessDigest->new->spin() runs at 04:55 rather than the 04:00 listed in the list config files? Are any pending digests processed or just left to the next day?

**Comment by dpc22** [comment]
Answering my own question: I'm pretty sure than:
`` 
my @timedigest = localtime $spool_digest->{time};`  
... 
if ($hh * 60 + $mm <= $now[2] * 60 + $now[1]
        and Time::Local::timelocal(0, @timedigest[1 .. 5]) <
        Time::Local::timelocal(0, $mm, $hh, @now[3 .. 5])) {
        return 1;
    }
```
Triggers whenever the current time is later than the nominated time in a day, and $spool_digest->{time} (the time of the first message in the digest?) is earlier that that time. 

Consequently it looks like it is safe to only run Sympa::Spindle::ProcessDigest->new->spin() once a day at 04:05 so long as list owners can't change $hh:$mm from 04:00.

**Comment by dpc22** [comment]
Slightly to my astonishment, the following works in edit_list.conf.
```
digest.hour             owner,privileged_owner          read
digest.minute           owner,privileged_owner          read
```
Well done Sympa authors.

**Comment by dpc22** [comment]
I am now running with the following modificates to Sympa 6.2.66, and will report back what happens:
```
--- sympa_msg.pl-BAK    2021-09-29 11:13:31.000000000 +0100
+++ sympa_msg.pl        2021-09-30 10:02:57.867533718 +0100
@@ -52,7 +52,7 @@
 
 ## Internal tuning
 # delay between each read of the digestqueue
-my $digestsleep = 5;
+my $digestsleep = 50;
 
 ## Init random engine
 srand(time());
@@ -237,7 +237,12 @@
     if (0 == $process->{generation}
         and $index_queuedigest++ >= $digestsleep) {
         $index_queuedigest = 0;
-        Sympa::Spindle::ProcessDigest->new->spin;
+
+        # Only run digests between 04:00 and 05:00
+        my @now = localtime time;
+        if ($now[2] == 4) {
+            Sympa::Spindle::ProcessDigest->new->spin;
+        }
     }
 
     $spindle->spin;

--- task_manager.pl-BAK 2021-09-29 11:13:31.000000000 +0100
+++ task_manager.pl     2021-09-30 10:03:04.218580903 +0100
@@ -150,7 +150,7 @@
     last if $spindle->{finish};
 
     # If the spool was empty, sleep for a while.
-    sleep 60;
+    sleep 1700;
 }
 
 # Purge grouped notifications
```

**Comment by racke** [comment]
Reading the digestqueue every 5 seconds seems ridiculous to me. Especially when the digest is send *once a day*.

**Comment by dpc22** [comment]
Its actually every 25 seconds (every 5th interaction, with a default 5 second delay between interations). 

I have arbitrary changed to once every 250 seconds (50*5), but only between  04:00 and 05:00.

ProcessDigest is actually quite fast when the digest spool is empty, but gets progressively worse through the course of the day if you have lots of lists and lots of messages.

**Comment by racke** [comment]
Ok every hour should be sufficient *and* the sleep time should be configurable for sure. I will check if I can make a PR for that.

**Comment by dpc22** [comment]
Can you spot the moment yesterday morning when I deployed these two changes?
["mid-term" load average for the last week](https://canvas.csi.cam.ac.uk/sympa_graphs/la.jpg)

sympa_msg.pl and task_manager.pl are now idle almost all of the time rather than taking turns to chase their own tails.  I'm pretty sure that the system would have coped with 7600 lists now that I understand what is going on, but this is much more comfortable. task_manager.pl really doesn't need to run more than a couple of times of hour. I personally don't need ProcessDigest more than once a day. 

Sympa 6.2.66 also seems to use much less memory than 6.2.64 for some reason:
[Memory use in bytes for the last week](https://canvas.csi.cam.ac.uk/sympa_graphs/mem.jpg)
I doubt that this is a consequence of my two changes, but I will take it!

**Comment by ikedas** [comment]
`sync_include` task should be run by duration not longer than `ttl` parameter of each list.  Default value of that parameter is 3600 (seconds) and the users can set it to be shorter.

That's why task_manager has to scan the lists frequently.  The other tasks may not be run so frequently.

**Comment by dpc22** [comment]
Fair point, although by default only listmaster can change the TTL on each list, /usr/share/sympa/default/edit_list.conf:
```
ttl                             owner,privileged_owner          hidden
```
and I don't plan on allowing anything less than the default 3600 seconds.
